File: android/app/src/test/java/com/nextpage/data/remote/addons/CuratedCatalogProviderTest.kt
package com.nextpage.data.remote.addons

import com.nextpage.data.remote.catalog.CatalogErrorCode
import com.nextpage.data.remote.catalog.CatalogException
import com.nextpage.data.remote.catalog.CompositeCatalogProvider
import com.nextpage.data.remote.catalog.CatalogProvider
import com.nextpage.data.remote.catalog.CatalogSourceInfo
import com.nextpage.data.remote.catalog.CatalogSourceKind
import com.nextpage.data.remote.catalog.BUILTIN_GUTENDEX
import com.nextpage.data.remote.catalog.BUILTIN_OPENLIBRARY
import com.nextpage.data.remote.catalog.DEBOUNCE_MS
import com.nextpage.data.remote.catalog.DEFAULT_PAGE_SIZE
import com.nextpage.data.remote.catalog.DETAIL_TTL_S
import com.nextpage.data.remote.catalog.MAX_PAGE_SIZE
import com.nextpage.data.remote.catalog.MIN_PAGE_SIZE
import com.nextpage.data.remote.catalog.OL_MIN_GAP_MS
import com.nextpage.data.remote.catalog.PAGE_TTL_S
import com.nextpage.data.remote.catalog.BUILTIN_OPENLIBRARY
import com.nextpage.data.remote.catalog.CatalogSourceKind.CURATED
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Test

/**
 * Curated first-party bundle parity suite. Mirrors desktop
 * curated-catalog-provider.test.ts: exactly six sources, every entry passes
 * ManifestValidator, browse-only behavior, stable not-routable errors.
 */
class CuratedCatalogProviderTest {

    private val curatedIds = listOf(
        "gutendex",
        "openlibrary",
        "standard-ebooks",
        "librivox",
        "wikisource",
        "faded-page"
    )

    private val curatedNames = listOf(
        "Gutendex",
        "Open Library",
        "Standard Ebooks",
        "LibriVox",
        "Wikisource",
        "Faded Page"
    )

    private fun providerFromBundle(): CuratedCatalogProvider {
        // Validate THE bundled asset (src/main/assets), not a test copy, so the
        // suite fails if the shipped bundle drifts.
        val candidates = listOf(
            java.io.File("src/main/assets/addons/curated.json"),
            java.io.File("app/src/main/assets/addons/curated.json")
        )
        val file = candidates.firstOrNull { it.exists() }
            ?: error("missing bundled asset addons/curated.json (tried ${candidates.joinToString()})")
        return CuratedCatalogProvider(file.readText())
    }

    @Test
    fun `bundle contains exactly the six first-party manifests in order`() {
        val provider = providerFromBundle()
        val sources = provider.listSources()
        assertEquals(6, sources.size)
        assertEquals(curatedIds.map { "builtin:$it" }, sources.map { it.sourceId })
        assertEquals(curatedNames, sources.map { it.name })
        assertEquals(List(6) { CURATED }, sources.map { it.kind })
    }

    @Test
    fun `every bundled entry passes ManifestValidator`() {
        // fromBundle validates each entry; a malformed bundle throws AddonFetchException.
        assertEquals(6, providerFromBundle().listSources().size)
    }

    @Test
    fun `search is browse-only empty page without I-O`() = kotlinx.coroutines.test.runTest {
        val provider = providerFromBundle()
        val page = provider.search("pride", 1)
        assertEquals(0, page.results.size)
        assertEquals(null, page.nextPage)
        assertEquals(0, page.totalCount)
    }

    @Test
    fun `getDetails fails stable not-routable`() = kotlinx.coroutines.test.runTest {
        val provider = providerFromBundle()
        try {
            provider.getDetails("standardebooks:some-book")
            fail("expected CatalogException")
        } catch (err: CatalogException) {
            assertEquals(CatalogErrorCode.NOT_FOUND, err.code)
        }
    }

    @Test
    fun `curated sources never shadow the built-ins in the composite`() =
        kotlinx.coroutines.test.runTest {
            val builtIn = object : CatalogProvider {
                override suspend fun search(query: String, page: Int): com.nextpage.data.remote.catalog.PagedResult {
                    error("not expected")
                }

                override suspend fun getDetails(id: String): com.nextpage.data.remote.catalog.CatalogBook {
                    error("not expected")
                }

                override fun resolveDownloadUrl(formats: Map<String, String>, preferEpub: Boolean): String = ""
                override fun listSources(): List<CatalogSourceInfo> = listOf(
                    CatalogSourceInfo(BUILTIN_GUTENDEX, "Gutendex", CatalogSourceKind.BUILTIN),
                    CatalogSourceInfo(BUILTIN_OPENLIBRARY, "Open Library", CatalogSourceKind.BUILTIN)
                )
            }
            val composite = CompositeCatalogProvider(listOf(builtIn, providerFromBundle()))
            val sources = composite.listSources()
            assertEquals(BUILTIN_GUTENDEX, sources.first().sourceId)
            assertEquals(CatalogSourceKind.BUILTIN, sources.first().kind)
            assertEquals(4, sources.count { it.kind == CatalogSourceKind.CURATED })
        }

    @Test
    fun `courtesy policy constants are untouched by the curated layer`() {
        assertEquals(350L, DEBOUNCE_MS)
        assertEquals(1_000L, OL_MIN_GAP_MS)
        assertEquals(24, DEFAULT_PAGE_SIZE)
        assertEquals(20, MIN_PAGE_SIZE)
        assertEquals(32, MAX_PAGE_SIZE)
        assertEquals(86_400L, PAGE_TTL_S)
        assertEquals(604_800L, DETAIL_TTL_S)
        assertTrue(true)
    }
}

File: desktop/src/test/unit/services/addons/curated-catalog-provider.test.ts
import { describe, expect, it } from 'vitest';
import { validateManifest } from '$lib/shared/services/addons/validateManifest';
import { CuratedCatalogProvider } from '$lib/shared/services/addons/CuratedCatalogProvider';
import curatedJson from '$lib/shared/services/addons/curated.json';
import { CompositeCatalogProvider } from '$lib/shared/services/catalog/CompositeCatalogProvider';
import { BUILTIN_GUTENDEX, BUILTIN_OPENLIBRARY } from '$lib/shared/services/catalog/CatalogProvider';
import type { CatalogProvider, PagedResult } from '$lib/shared/services/catalog/CatalogProvider';
import {
  DETAIL_TTL_S,
  PAGE_TTL_S,
} from '$lib/shared/services/catalog/DiscoverCache';
import {
  DEBOUNCE_MS,
  DEFAULT_PAGE_SIZE,
  MAX_PAGE_SIZE,
  MIN_PAGE_SIZE,
  OL_MIN_GAP_MS,
} from '$lib/shared/services/catalog/policy';

const CURATED_IDS = [
  'gutendex',
  'openlibrary',
  'standard-ebooks',
  'librivox',
  'wikisource',
  'faded-page',
];

const CURATED_NAMES = [
  'Gutendex',
  'Open Library',
  'Standard Ebooks',
  'LibriVox',
  'Wikisource',
  'Faded Page',
];

function entryBytes(entry: unknown): Uint8Array {
  return new TextEncoder().encode(JSON.stringify(entry));
}

describe('curated bundle', () => {
  it('contains exactly the six first-party manifests in order', () => {
    const entries = (curatedJson as { addons: unknown[] }).addons;
    expect(entries).toHaveLength(6);
    expect(entries.map((e) => (e as { id: string }).id)).toEqual(CURATED_IDS);
    expect(entries.map((e) => (e as { name: string }).name)).toEqual(CURATED_NAMES);
  });

  it('every entry passes the shared manifest validation', () => {
    const entries = (curatedJson as { addons: unknown[] }).addons;
    for (const entry of entries) {
      expect(() => validateManifest(entryBytes(entry), 'application/json')).not.toThrow();
    }
  });
});

describe('CuratedCatalogProvider', () => {
  it('listSources() returns six curated sources derived from manifest ids', () => {
    const provider = new CuratedCatalogProvider();
    expect(provider.listSources()).toEqual(
      CURATED_IDS.map((id, i) => ({
        sourceId: `builtin:${id}`,
        name: CURATED_NAMES[i],
        kind: 'curated' as const,
      })),
    );
  });

  it('search is browse-only: empty page, no I/O, no throw', async () => {
    const provider = new CuratedCatalogProvider();
    const page = await provider.search('pride', 1);
    expect(page).toEqual({ results: [], nextPage: null, totalCount: 0 });
  });

  it('getDetails fails stable not-routable (NOT_FOUND, no crash)', async () => {
    const provider = new CuratedCatalogProvider();
    const err = await provider.getDetails('standardebooks:some-book').catch((e: unknown) => e);
    expect((err as { code: string }).code).toBe('NOT_FOUND');
  });
});

describe('composite integration', () => {
  class EmptyProvider implements CatalogProvider {
    async search(): Promise<PagedResult> {
      throw new Error('not expected');
    }
    async getDetails(id: string): Promise<never> {
      throw new Error(id);
    }
    resolveDownloadUrl(): string {
      return '';
    }
    listSources() {
      return [
        { sourceId: BUILTIN_GUTENDEX, name: 'Gutendex', kind: 'builtin' as const },
        { sourceId: BUILTIN_OPENLIBRARY, name: 'Open Library', kind: 'builtin' as const },
      ];
    }
  }

  it('curated sources never shadow the built-ins (dedupe keeps first/builtin kind)', () => {
    const provider = new CompositeCatalogProvider([new EmptyProvider(), new CuratedCatalogProvider()]);
    const sources = provider.listSources();
    expect(sources[0]).toEqual({
      sourceId: 'builtin:gutendex',
      name: 'Gutendex',
      kind: 'builtin',
    });
    expect(sources.filter((s) => s.kind === 'curated')).toHaveLength(4);
  });

  it('courtesy policy constants are untouched by the curated layer', () => {
    expect(DEBOUNCE_MS).toBe(350);
    expect(OL_MIN_GAP_MS).toBe(1_000);
    expect(DEFAULT_PAGE_SIZE).toBe(24);
    expect(MIN_PAGE_SIZE).toBe(20);
    expect(MAX_PAGE_SIZE).toBe(32);
    expect(PAGE_TTL_S).toBe(86_400);
    expect(DETAIL_TTL_S).toBe(604_800);
  });
});
